### PR TITLE
Add documentation about deployment

### PR DESCRIPTION
**Pull request by fabcor-maxiv**
Preview: https://mxcubeweb--1193.org.readthedocs.build/en/1193/dev/deployment.html

### CODE REVIEW

File: docs/source/dev/deployment.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    poetry export --format=requirements.txt > requirements.txt
+    ```
+
+1. Move the `requirements.txt` file to the deployment target system
+
+1. On the deployment system, let pip install {{project}} via the exported requirements file
+
+    1. Install the dependencies
+        ```shell
+        python -m pip install --no-deps mxcubeweb
```
Comment (axelboc) [inline docs/source/dev/deployment.md]: Did you mean to keep this line?

File: docs/source/dev/deployment.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# Deployment
+
+{{ project }} uses [*Poetry*](https://python-poetry.org/) as development workflow tool.
+One of Poetry's feature is support for a so-called *lockfile*.
```
Comment (axelboc) [inline docs/source/dev/deployment.md]:
```suggestion
One of Poetry's features is support for a so-called *lockfile*.
```

**Review by axelboc — APPROVED** [review]

**Comment by fabcor-maxiv** [comment]
@axelboc What about the JavaScript part? Anything that could be mentioned? What should one pay attention to when deploying for production regarding the JavaScript part? Some special flags to set when building? Anything like that?

File: docs/source/dev/deployment.md (hunk 1 above)
Comment (fabcor-maxiv) [reply docs/source/dev/deployment.md]: No : D

Removed, thanks!

File: docs/source/dev/deployment.md (hunk 2 above)
Comment (fabcor-maxiv) [reply docs/source/dev/deployment.md]: Thanks : )

**Comment by axelboc** [comment]
> @axelboc What about the JavaScript part? Anything that could be mentioned? What should one pay attention to when deploying for production regarding the JavaScript part? Some special flags to set when building? Anything like that?

Well, I guess you could copy the steps from the README: 1) install the front-end dependencies with `pnpm --prefix ui install`; 2) build the front-end with `pnpm --prefix ui build` 3) point `maxcube-server` to where the built files are located (`ui/build`).

There's a `.env` file in the `ui` folder but it's meant to be removed. I don't see anything else but @marcus-oscarsson may have a better idea.

File: docs/source/dev/deployment.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+to install and deploy in a production environment.
+Some might prefer to use a dedicated installer tool, like pip for such a task.
+
+But Poetry's lockfile format is not standardized, its content is specific to Poetry.
+So this file is not immediately usable by other installers and tools.
+
+To circumvent this, one could use a command like `poetry export`
+to generate a `requirements.txt` file compatible with pip (and some other tools).
+With a series of commands like the following,
+it should be possible to approximate the behaviour of installing with Poetry's lockfile.
```
Comment (marcus-oscarsson) [inline docs/source/dev/deployment.md:41]: It is possible to ... ?

**Comment by marcus-oscarsson** [comment]
Yes the .env file is no longer used and can be removed.

**Comment by marcus-oscarsson** [comment]
I'm wondering if we should not provide a pre-built UI and bundle that with the python package.

**Comment by marcus-oscarsson** [comment]
Should we make sure that there is a `requirements.txt` that is up to date in the repository if we recommend using `pip` for deployment or de we leave that up to each and everyone to decide ?

**Comment by marcus-oscarsson** [comment]
Thanks, looks good to me !

**Comment by axelboc** [comment]
> I'm wondering if we should not provide a pre-built UI and bundle that with the python package.

If there are no longer any environment variables and no other build-time configuration, then for sure! Not having to build the front-end would definitely make deployment simpler.

File: docs/source/dev/deployment.md (hunk 3 above)
Comment (fabcor-maxiv) [reply docs/source/dev/deployment.md:41]: Should be. My goal with this document is to give some pointers, and every facility can build their workflow for themselves according to their own infrastructure. There is no way we can come up with one deployment method that works for all facilities.

But yes, I intend to check this again, make sure it works at least with the default out of the box MXCuBE-Web (with the mockups).

**Comment by fabcor-maxiv** [comment]
> I'm wondering if we should not provide a pre-built UI and bundle that with the python package.

I can not think of a way to do this that would not be a pain, but I have not thought about it very long. If someone comes up with a good solution, then that would be great of course.


> Should we make sure that there is a requirements.txt that is up to date in the repository if we recommend using pip for deployment or de we leave that up to each and everyone to decide ?

Yes, I have already experimented with this a bit. I think with pre-commit hooks it can be done without too much pain and suffering inflicted on contributors. But I would do it in a separate effort (pull request).


> There's a .env file in the ui folder but it's meant to be removed. I don't see anything else but @marcus-oscarsson may have a better idea.

> Yes the .env file is no longer used and can be removed.

No idea about this. Feel free to do it in a separate pull request, I guess.

File: docs/source/dev/deployment.md (hunk 3 above)
Comment (marcus-oscarsson) [reply docs/source/dev/deployment.md:41]: I see,  using "should be" sounds quite theoretical/hypothetical and we would at-least assume that we have documented something that works. The aim would not be to come up with a complete deployment procedure that works for everyone but at least outline some good practices and make sure that everyone can use them if they would like to.

**Comment by marcus-oscarsson** [comment]
I see, I think that if we agree that we should use pip for deployment then we should IMO make sure that there is either a script that generates the `requirements.txt` and performs the `pip install` or automatically create the `requirements.txt`

**Comment by marcus-oscarsson** [comment]
I think we can come up with a way to build and bundle the UI with the python package in someway.

**Comment by fabcor-maxiv** [comment]
@marcus-oscarsson @axelboc dedicated issues to discuss the tangential topics:
* https://github.com/mxcube/mxcubeweb/issues/1219
* https://github.com/mxcube/mxcubeweb/issues/1220

File: docs/source/dev/deployment.md (hunk 3 above)
Comment (fabcor-maxiv) [reply docs/source/dev/deployment.md:41]: I tested again just now, and it works as far as I can tell.

**Comment by fabcor-maxiv** [comment]
I have thought about adding a section about the Javascript UI part that would not be a complete copy/paste of the other page, and I can not come up with anything. So I rather leave this out. We can always add something when we have something concrete to write that is actual additional information.

So I guess this is ready to be merged from my point of view.

**Pull request merged**